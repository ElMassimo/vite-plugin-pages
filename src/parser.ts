import { promises as fs } from 'fs'
import { extname } from 'path'
import JSON5 from 'json5'
import matter from 'gray-matter'

import { ResolvedOptions } from './types'
import type { SFCDescriptor } from '@vue/compiler-sfc'

const engines = { json5: JSON5.parse.bind(JSON5) }

async function parseSFC(code: string): Promise<SFCDescriptor> {
  try {
    const { parse } = await import('@vue/compiler-sfc')
    return parse(code, {
      pad: 'space',
    }).descriptor
  } catch {
    throw new Error('[vite-plugin-pages] Vue3\'s "@vue/compiler-sfc" is required.')
  }
}

export async function parseRouteData(filePath: string, options: ResolvedOptions) {
  const content = await fs.readFile(filePath, 'utf8')
  return extname(filePath) === '.vue'
    ? await parseCustomBlock(filePath, content, options)
    : parseMarkdownFile(filePath, content)
}

function parseFrontmatter(content: string, language?: string) {
  return matter(content, { language, engines }).data
}

export function parseMarkdownFile(filePath: string, content: string) {
  try {
    return parseFrontmatter(content)
  } catch (err: any) {
    throw new Error(`Invalid frontmatter for ${filePath}\n${err.message}`)
  }
}

export async function parseCustomBlock(filePath: string, content: string, options: ResolvedOptions) {
  const parsed = await parseSFC(content)
  const block = parsed.customBlocks.find(b => b.type === 'route')
  if (!block) return undefined
  const language = block.lang || options.routeBlockLang
  try {
    return {
      ...parsed.template?.attrs,
      ...parseFrontmatter(`---\n${block.content}\n---`, language),
    }
  } catch (err: any) {
    throw new Error(`Invalid ${language} format of <${block.type}> content in ${filePath}\n${err.message}`)
  }
}
