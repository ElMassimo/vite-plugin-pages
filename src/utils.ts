import { resolve, basename } from 'path'
import Debug from 'debug'
import deepEqual from 'deep-equal'
import type { Nullable, Arrayable } from '@antfu/utils'
import { ResolvedOptions, Route } from './types'
import { parseRouteData } from './parser'
import { MODULE_ID_VIRTUAL } from './constants'
import type { ViteDevServer } from 'vite'
import type { OutputBundle } from 'rollup'

/**
 * Convert `Arrayable<T>` to `Array<T>`
 */
export function toArray<T>(array?: Nullable<Arrayable<T>>): Array<T> {
  array = array || []
  if (Array.isArray(array))
    return array
  return [array]
}

/**
 * Replace backslash to slash
 */
export function slash(str: string) {
  return str.replace(/\\/g, '/')
}

export const routeBlockCache = new Map<string, Record<string, any>>()

export function extensionsToGlob(extensions: string[]) {
  return extensions.length > 1 ? `{${extensions.join(',')}}` : extensions[0] || ''
}

/**
 * Clear undefined fields from an object. It mutates the object
 */
export function clearUndefined<T extends object>(obj: T): T {
  // @ts-expect-error
  Object.keys(obj).forEach((key: string) => (obj[key] === undefined ? delete obj[key] : {}))
  return obj
}

function isPagesDir(path: string, options: ResolvedOptions) {
  for (const page of options.pagesDir) {
    const dirPath = slash(resolve(options.root, page.dir))
    if (path.startsWith(dirPath)) return true
  }
  return false
}

export function isTarget(path: string, options: ResolvedOptions) {
  return isPagesDir(path, options) && options.extensionsRE.test(path)
}

export function supportsCustomBlock(path: string, options: ResolvedOptions) {
  return isTarget(path, options) && !options.react
}

export const debug = {
  hmr: Debug('vite-plugin-pages:hmr'),
  parser: Debug('vite-plugin-pages:parser'),
  gen: Debug('vite-plugin-pages:gen'),
  options: Debug('vite-plugin-pages:options'),
  cache: Debug('vite-plugin-pages:cache'),
  pages: Debug('vite-plugin-pages:pages'),
}

const dynamicRouteRE = /^\[.+\]$/
export const nuxtDynamicRouteRE = /^_[\s\S]*$/

export function isDynamicRoute(routePath: string, nuxtStyle: Boolean = false) {
  return nuxtStyle
    ? nuxtDynamicRouteRE.test(routePath)
    : dynamicRouteRE.test(routePath)
}

export function isCatchAllRoute(routePath: string, nuxtStyle: Boolean = false) {
  return nuxtStyle
    ? /^_$/.test(routePath)
    : /^\[\.{3}/.test(routePath)
}

export function resolveImportMode(
  filepath: string,
  options: ResolvedOptions,
) {
  const mode = options.importMode
  if (typeof mode === 'function')
    return mode(filepath)

  for (const pageDir of options.pagesDir) {
    if (
      options.syncIndex
      && pageDir.baseRoute === ''
      && filepath === `/${pageDir.dir}/index.vue`
    )
      return 'sync'
  }
  return mode
}

export function pathToName(filepath: string) {
  return filepath.replace(/[_.\-\\/]/g, '_').replace(/[[:\]()]/g, '$')
}

export function findRouteByFilename(routes: Route[], filename: string): Route | null {
  let result = null
  for (const route of routes) {
    if (filename.endsWith(route.component))
      result = route

    if (!result && route.children)
      result = findRouteByFilename(route.children, filename)

    if (result) return result
  }
  return null
}

export async function getRouteBlock(path: string, options: ResolvedOptions) {
  if (!isTarget(path, options) || options.react) return null

  let result
  try {
    result = await parseRouteData(path, options)
  } catch (error: any) {
    if (!options.server) throw error
    options.server.config.logger.error(error.message, { timestamp: true, error })
    options.server.ws.send({ type: 'error', err: error })
  }
  if (!result) return null

  debug.parser('%s: %O', path, result)
  routeBlockCache.set(slash(path), result)

  return result
}

export function getPagesVirtualModule(server: ViteDevServer) {
  const { moduleGraph } = server
  const module = moduleGraph.getModuleById(MODULE_ID_VIRTUAL)
  if (module) {
    moduleGraph.invalidateModule(module)
    return module
  }
  return null
}

export function replaceSquareBrackets(bundle: OutputBundle) {
  const files = Object.keys(bundle).map(i => basename(i))
  for (const chunk of Object.values(bundle)) {
    chunk.fileName = chunk.fileName.replace(/(\[|\])/g, '_')
    if (chunk.type === 'chunk') {
      for (const file of files)
        chunk.code = chunk.code.replace(file, file.replace(/(\[|\])/g, '_'))
    }
  }
}

export async function checkRouteBlockChanges(filePath: string, options: ResolvedOptions) {
  debug.cache(routeBlockCache)
  const oldRouteBlock = routeBlockCache.get(filePath)
  const routeBlock = await getRouteBlock(filePath, options)

  debug.hmr('%s old: %O', filePath, oldRouteBlock)
  debug.hmr('%s new: %O', filePath, routeBlock)

  return {
    changed: !deepEqual(oldRouteBlock, routeBlock),
    needsReload: !deepEqual(oldRouteBlock?.route, routeBlock?.route)
      || !deepEqual(oldRouteBlock?.templateAttrs, routeBlock?.templateAttrs),
  }
}
