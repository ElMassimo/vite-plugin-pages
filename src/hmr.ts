import { ViteDevServer } from 'vite'
import { getPagesVirtualModule, checkRouteBlockChanges, isTarget, supportsCustomBlock, debug, slash } from './utils'
import { removePage, addPage, updatePage } from './pages'
import { ResolvedOptions, ResolvedPages } from './types'

export function handleHMR(server: ViteDevServer, pages: ResolvedPages, options: ResolvedOptions, clearRoutes: () => void) {
  const { ws, watcher, moduleGraph } = server

  function fullReload() {
    // invalidate module
    getPagesVirtualModule(server)
    clearRoutes()
    ws.send({
      type: 'full-reload',
    })
  }

  watcher.on('add', async(file) => {
    const path = slash(file)
    if (isTarget(path, options)) {
      await addPage(pages, path, options)
      debug.hmr('add', path)
      fullReload()
    }
  })
  watcher.on('unlink', (file) => {
    const path = slash(file)
    if (isTarget(path, options)) {
      removePage(pages, path)
      debug.hmr('remove', path)
      fullReload()
    }
  })
  watcher.on('change', async(file) => {
    const path = slash(file)
    if (supportsCustomBlock(file, options)) {
      const { changed, needsReload } = await checkRouteBlockChanges(path, options)
      if (changed) {
        moduleGraph.getModulesByFile(file)?.forEach(mod => moduleGraph.invalidateModule(mod))
        updatePage(pages, path)
        debug.hmr('change', path)
        if (needsReload) fullReload()
      }
    }
  })
}
