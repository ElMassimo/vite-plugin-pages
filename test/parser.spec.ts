import { resolve } from 'path'
import { parseRouteData, parseMarkdownFile } from '../src/parser'
import { resolveOptions } from '../src/options'

const options = resolveOptions({})

describe('Parser', () => {
  test('json5 block', async() => {
    const path = resolve('./test/assets/pages/blog/[id].vue')
    const parsedCustomBlock = await parseRouteData(path, options)
    expect(parsedCustomBlock).toMatchSnapshot('custom block')
  })

  test('empty block', async() => {
    const path = resolve('./test/assets/pages/about.vue')
    const parsedCustomBlock = await parseRouteData(path, options)
    expect(parsedCustomBlock).toEqual(undefined)
  })

  test('yaml block', async() => {
    const path = resolve('./test/assets/pages/components.vue')
    const parsedCustomBlock = await parseRouteData(path, options)
    expect(parsedCustomBlock).toMatchSnapshot('yaml block')
  })

  test('markdown file', async() => {
    const content = await parseMarkdownFile('example.md', '---\ntitle: About\n---')
    expect(content).toEqual({ meta: { title: 'About' } })
  })
})
