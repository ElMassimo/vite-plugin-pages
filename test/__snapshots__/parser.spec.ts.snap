// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser json5 block: custom block 1`] = `
Object {
  "meta": Object {
    "requiresAuth": false,
  },
  "name": "name-override",
}
`;

exports[`Parser yaml block: yaml block 1`] = `
Object {
  "meta": Object {
    "lang": "es",
  },
}
`;
